src/sources/ping/eventfd.rs

```diff
@@ -0,0 +1,190 @@
+//! Eventfd based implementation of the ping event source.
+//!
+//! # Implementation notes
+//!
+//! The eventfd is a much lighter signalling mechanism provided by the Linux
+//! kernel. Rather than write an arbitrary sequence of bytes, it only has a
+//! 64-bit counter.
+//!
+//! To avoid closing the eventfd early, we duplicate it in `make_ping()` with
+//! `dup()`. The returned fd can then be closed when all the senders are dropped
+//! without invalidating the source's fd.
+//!
+//! To differentiate between regular ping events and close ping events, we add 2
+//! to the counter for regular events and 1 for close events. In the source we
+//! can then check the LSB and if it's set, we know it was a close event. This
+//! only works if a close event never fires more than once.
+
+use std::{os::unix::io::RawFd, sync::Arc};
+
+use nix::sys::eventfd::{eventfd, EfdFlags};
+use nix::unistd::{close, dup, read, write};
+
+use super::PingError;
+use crate::{
+    generic::Generic, EventSource, Interest, Mode, Poll, PostAction, Readiness, Token, TokenFactory,
+};
+
+// These are not bitfields! They are increments to add to the eventfd counter.
+// Since the fd can only be closed once, we can effectively use the
+// INCREMENT_CLOSE value as a bitmask when checking.
+const INCREMENT_PING: u64 = 0x2;
+const INCREMENT_CLOSE: u64 = 0x1;
+
+#[inline]
+pub fn make_ping() -> std::io::Result<(Ping, PingSource)> {
+    let read = eventfd(0, EfdFlags::EFD_CLOEXEC | EfdFlags::EFD_NONBLOCK)?;
+    // We only have one fd for the eventfd. If the sending end closes it when
+    // all copies are dropped, the receiving end will be closed as well. We can
+    // avoid this by duplicating the FD.
+    let write = dup(read)?;
```

Comment:
If this fails we are leaking a file descriptor

Reply:
Can we use `Arc` instead of `dup` here? I don't see a reason to dup it. We could have `Weak<Fd>` pointers in senders and check in receiver ref count and drop fd it goes to zero.

Reply:
> We could have Weak<Fd> pointers in senders and check in receiver ref count and drop fd it goes to zero.

This would work *if* we can be sure the weak count (ie. number of senders) will never go to zero *and then* come back up again. Can anyone think of a situation where that might happen? (I can't.)

Reply:
I'm not quite sure I get how your suggestion would work out @kchibisov

Do you mean checking the refcount every time the source is woken up, and drop the source if it reaches zero? It seems to me that it implies waking up the source every time a `Ping` is dropped, and how would we be able to distinguish that from a regular `ping()`? Dropping a `Ping` handle would cause spurious events on the event source, no?

Reply:
Yeah I just tried this approach and hit this exact problem. Spurious events aren't such a big deal for me (you get a lot of them with the async executor for example, depending on what's in there), but yes, there's that. And then there's the "what kind of wakeup was this" issue, which means we'd still need the `FlagOnDrop` trick, which means two levels of `Arc` (on for sharing the fd between senders and source, and one for sharing it between just the senders).

Reply:
@vberger I've also realized that it's a **problem** later on. In general I don't want to have lots of fd's. The issue is that identifying that certain ping should dispatch callback, but some other don't is complicated, and I'm not sure that we can solve it without wakeups.

I think what we can do is to pass some value for every ping, but when the drop is performed the value should be zero. So if we woken up and the value is zero then we shouldn't dispatch callback, and if ref count went to zero it means that drop should be performed. And every non zero is drained on read, so it should be fine.

Reply:
@detly have you tried writing `1` for regular pings and `0` for drop ones? So you could check if the value is `zero` in wakeup => no callback call, and it ref count went to zero as well it means that drop should be done?

I think it could work without extra flags, just not sure if passing zero is allowed.

Reply:
@kchibisov Alas passing 0 is not allowed, or rather it is, but it won't trigger an event on the `eventfd`.

Reply:
Oh, there *is* a semaphore mode for the eventfd. Let me take a look at that.

Reply:
Nope, it's more about semaphore *events* not ownership, I don't think it helps us.

Reply:
At any rate, my recent push fixes the fd leak, but I'm happy to rework the sender management if there's a better way.

Reply:
If you're insisting on using `dup` and not pass it around by arc, be aware that `dup` discards flags, like `EFD_CLOEXEC`, so you should set them manually. Not sure about other apis nix provide.

Reply:
Good catch, I will fix that too.

I'm not so much insisting on it, more that I can't really get a recount approach to work without introducing some other problem (loop wakeups, race conditions, or API ugliness are the main ones).

Reply:
I mean you don't need ref count to work, you can use your approach, but use arc instead of dup. The only change you'll add is remove close from client, and check refcount before dropping.

Reply:
> The only change you'll add is remove close from client, and check refcount before dropping.

What happens if the source is dropped before the senders?

Reply:
The `Weak` will tell you that?

Reply:
> The Weak will tell you that? It's like weak for a reason...

Right, so now we're back to - how do you unambiguously signal to the source that a sender has been dropped? Dropping the last `Weak` ref won't trigger a drop, because the source needs to hold on to the original `Arc`.

Reply:
Maybe wrap the `Weak` in another `Arc<FlagOnDrop>` that's shared amongst all the senders? That could work...

Reply:
Ah, yes, I think I have something now.

Reply:
Okay. Okay. I think this last push might satisfy everyone. Thanks for your persistence @kchibisov, this really should be a lot safer. I added another test to check the case of multiple senders and dropping them in between dispatches.

Reply:
Do I understand that the only point of using `Weak` here is so that the FD is closed as soon as the `PingSource` is dropped, rather than once the `PingSource` and all `Ping`s are dropped?

Reply:
Sort of. It was partly that, and also so that simply holding on to a `Ping` did not keep the `PingSource`'s underlying eventfd alive.

I see your point though, the implementation would be simpler if it was `Arc`s all the way.

Reply:
Latest version has full `Arc`s all the way through. I've left the intermediate commits in in case you want to compare, but I'll squash them when we're happy with it.

Reply:
I don't think that's a blocking point though, that was just to make sure I understand the point.

Reply:
I think I prefer the full-Arc version, it *is* simpler in general.